### PR TITLE
regif documented

### PR DESCRIPTION
**Pull request by jijingg**
here to pay my debts 😆，register interface build tools document done!

### CODE REVIEW

File: source/SpinalHDL/Libraries/regIf.rst [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+      io.interrupt := (tx_int_mask(0) && tx_int_state(0)  ||
+        rx_int_mask(0) && rx_int_state(0) ||
+        frame_int_mask(0) && frame_int_state(0))
+
+   }
+
+this is a very tedious and repetitive work, a better way is creat a Factory fucntion by Macros auto complet those work instead manully creat them.
+now th InterruptFactory can do that.
+    
+Easy Way creat interruption:
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:169]:
```suggestion
this is a very tedious and repetitive work, a better way is to use the "factory" paradigm to auto-generate the documentation for each signal.

now th InterruptFactory can do that.
    
Easy Way creat interruption:
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 2]
```diff
@@ -0,0 +1,208 @@
+ 
+RegIf
+============
+Register Interface Builder
+
+- Automatic address and fields allocate, and conflict detection
+- 28 Register Access types(Covering 25 types of UVM)
+- Auto Document generate
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:8]:
```suggestion
- Automatic address, fields allocation and conflict detection
- 28 Register Access types(Covering the 25 types defined by the UVM standard)
- Automatic documentation generation
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+
+Automatic allocate
+------------------
+
+Automatic address allocate
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:13]:
```suggestion
Automatic address allocation
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+    val M_REGn1 = busif.newReg(doc="REGn1")
+
+    busif.accept(HtmlGenerator("regif.html", "AP"))
+    // busif.accept(CHeaderGenerator("header.h", "AP"))
+    // busif.accept(JsonGenerator("regif.json"))
+  }
+
+.. image:: /asset/image/regif/reg-auto-allocate.gif
+
+Automatic fileds allocate
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:36]:
```suggestion
Automatic fileds allocation
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+  val M_REG1  = busif.newReg(doc="REG1")
+  val r1fd0 = M_REG1.field(16 bits, RW, doc="fields 1")
+  val r1fd2 = M_REG1.field(offset=10, 2 bits, RW, doc="fields 1")
+    ...
+  cause Exception
+
+28 Access Type
+--------------
+  
+Most of these come from UVM
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:69]:
```suggestion
Most of these come from UVM specification
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+WOS         w: sets all bits, r: error                                                      UVM
+W1          w: first one after hard reset is as-is, other w have no effects, r: no effect   UVM
+WO1         w: first one after hard reset is as-is, other w have no effects, r: error       UVM
+NA          w: reserved, r: reserved                                                        New
+W1P         w: 1/0 pulse/no effect on matching bit, r: no effect                            New
+W0P         w: 0/1 pulse/no effect on matching bit, r: no effect                            New
+==========  =============================================================================   ====
+
+Auto generate document
+----------------------
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:105]:
```suggestion
Automatic documentation generation
----------------------------------
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 7]
```diff
@@ -0,0 +1,208 @@
...
+
+      busif.accept(HtmlGenerator("interrupt.html", "AP"))
+    }
+
+.. image:: /asset/image/regif/easy-intr.png
+
+Developers Area
+===============
+
+You Extend document Type based on Trait BusIfVistor
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:191]:
```suggestion
You can add your document Type by extending the `BusIfVistor` Trait 
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 8]
```diff
@@ -0,0 +1,208 @@
...
+    ...
+  cause Exception
+  val M_REG1  = busif.newReg(doc="REG1")
+  val r1fd0 = M_REG1.field(16 bits, RW, doc="fields 1")
+  val r1fd2 = M_REG1.field(offset=10, 2 bits, RW, doc="fields 1")
+    ...
+  cause Exception
+
+28 Access Type
+--------------
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:67]:
```suggestion
28 Access Types
---------------
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+RegIf
+============
+Register Interface Builder
+
+- Automatic address and fields allocate, and conflict detection
+- 28 Register Access types(Covering 25 types of UVM)
+- Auto Document generate
+
+Automatic allocate
+------------------
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:11]:
```suggestion
Automatic allocation
--------------------
```

File: source/SpinalHDL/Libraries/regIf.rst [hunk 10]
```diff
@@ -0,0 +1,208 @@
+ 
+RegIf
+============
```
Comment (wifasoi) [inline source/SpinalHDL/Libraries/regIf.rst:3]:
```suggestion
RegIf
=====
```

**Review by wifasoi — CHANGES_REQUESTED** [review]
I rewritten some part.
You can add my correction, or just ignore them.

Anyway, thanks for the docs;
Now your doc tax is payed and you can sleep peacefully... FOR NOW! *\***evil laughter***\*

**Comment by wifasoi** [comment]
Don't worry about the cicd.. for now.
I didn't have the time to fix the dam thing yet :D

**Comment by jijingg** [comment]
> I rewritten some part.

thank you very much for your careful revision, ❤️ 

> you can sleep peacefully... FOR NOW

 haha..  you'r right

**Comment by jijingg** [comment]
> You can add my correction

done ,  thanks

**Pull request merged**